diff --git a/ports/rp2/main.c b/ports/rp2/main.c
index d976464de..d43d8fe96 100644
--- a/ports/rp2/main.c
+++ b/ports/rp2/main.c
@@ -49,6 +49,7 @@
 #include "pico/unique_id.h"
 #include "hardware/rtc.h"
 #include "hardware/structs/rosc.h"
+#include "hardware/adc.h"
 #if MICROPY_PY_LWIP
 #include "lwip/init.h"
 #include "lwip/apps/mdns.h"
@@ -68,6 +69,8 @@
 extern uint8_t __StackTop, __StackBottom;
 static char gc_heap[MICROPY_GC_HEAP_SIZE];
 
+extern uint16_t runtime_wakeup_vsys_voltage;
+
 // Embed version info in the binary in machine readable form
 bi_decl(bi_program_version_string(MICROPY_GIT_TAG));
 
@@ -126,6 +129,26 @@ int main(int argc, char **argv) {
     #endif
     #endif

+    // Initialise GP25 as an output so VSYS can be read
+    gpio_init(25);
+    gpio_set_dir(25, true);
+    gpio_put(25, true);
+
+    // Initialise the ADC and select the input for reading VSYS
+    adc_init();
+    adc_gpio_init(29);
+    adc_select_input(3);
+
+    // Read the VSYS voltage 16 times to produce a stable value
+    runtime_wakeup_adc_value = 0u;
+    for(uint8_t i = 0u; i < 16u; i++) {
+        runtime_wakeup_adc_value += adc_read();
+    }
+
+    // Cleanup GP25 and 29
+    gpio_init(29);
+    gpio_init(25);
+
     #if MICROPY_PY_NETWORK_CYW43
     {
         cyw43_init(&cyw43_state);
